src/components/connectAdvanced.js

```diff
@@ -118,146 +82,130 @@ export default function connectAdvanced(
 
-      runUpdater(callback = noop) {
-        if (this.isUnmounted) {
-          return
-        }
+        const shouldUpdate = childPropsChanged || hasError;
+        return shouldUpdate;
 
-        this.setState(prevState => prevState.updater(this.props, prevState), callback)
       }
 
-      initSubscription() {
-        if (!shouldHandleStateChanges) return
-
-        // parentSub's source should match where store came from: props vs. context. A component
-        // connected to the store via props shouldn't use subscription from context, or vice versa.
-        const parentSub = (this.propsMode ? this.props : this.context)[subscriptionKey]
-        this.subscription = new Subscription(this.store, parentSub, this.onStateChange.bind(this))
-
-        // `notifyNestedSubs` is duplicated to handle the case where the component is  unmounted in
-        // the middle of the notification loop, where `this.subscription` will then be null. An
-        // extra null check every change can be avoided by copying the method onto `this` and then
-        // replacing it with a no-op on unmount. This can probably be avoided if Subscription's
-        // listeners logic is changed to not call listeners that have been unsubscribed in the
-        // middle of the notification loop.
-        this.notifyNestedSubs = this.subscription.notifyNestedSubs.bind(this.subscription)
-      }
+      render() {
+        if(this.state.error) {
+          throw this.state.error;
+        }
 
-      onStateChange() {
-        this.runUpdater(this.notifyNestedSubs)
+        return <WrappedComponent {...this.state.childProps} />
       }
+    }
 
-      isSubscribed() {
-        return Boolean(this.subscription) && this.subscription.isSubscribed()
-      }
+    class Connect extends OuterBaseComponent {
+      constructor(props) {
+        super(props)
 
+        this.renderInner = this.renderInner.bind(this);
+      }
+/*
       addExtraProps(props) {
-        if (!withRef && !renderCountProp && !(this.propsMode && this.subscription)) return props
+          if (!withRef && !renderCountProp) return props;
+
         // make a shallow copy so that fields added don't leak to the original selector.
         // this is especially important for 'ref' since that's a reference back to the component
         // instance. a singleton memoized selector would then be holding a reference to the
         // instance, preventing the instance from being garbage collected, and that would be bad
         const withExtras = { ...props }
-        if (withRef) withExtras.ref = this.setWrappedInstance
+        //if (withRef) withExtras.ref = this.setWrappedInstance
         if (renderCountProp) withExtras[renderCountProp] = this.renderCount++
-        if (this.propsMode && this.subscription) withExtras[subscriptionKey] = this.subscription
+
         return withExtras
       }
+*/
+
+      renderInner(providerValue) {
+          const {storeState, store} = providerValue;
```

Comment:
here, I would rather do the calculation of derived props and memoize them, then the inner component need only be the WrappedComponent or a PureComponent that wraps it, something like:

```js
const Inner = connectOptions.pure ?
  class Inner extends PureComponent {
    render() {
      return <WrappedComponent {...this.props} />
    }
  } : WrappedComponent
```

then renderInner can be something like:

```js
  renderInner({ storeState, store }) {
    const props = this.props
    const derivedProps = this.memoizedProps(storeState, props)
    return <Inner {...props} {...derivedProps} />
  }
```

Our memoizer can be initialized in the constructor with `this.memoizedProps = this.makeMemoizer()`

```js
  makeMemoizer() {
    let lastProps
    let lastState
    let lastDerivedProps
    let called = false
    return (state, props) => {
      if (called) {
        const sameProps = connectOptions.pure && shallowEqual(lastProps, props)
        const sameState = lastState === state
        if (sameProps && sameState) {
          return lastDerivedProps
        }
      }
      called = true
      lastProps = props
      lastState = state
      const nextProps = this.state.childPropsSelector(state, props)
      if (shallowEqual(lastDerivedProps, nextProps)) {
        return lastDerivedProps
      }
      lastDerivedProps = nextProps
      return lastDerivedProps
    }
  }
```

In this way, we don't need `getDerivedStateFromProps` at all

Reply:
Let's try that approach as a second PR, for comparison purposes.